src/visit.ts

```diff
@@ -285,27 +287,38 @@ export function visitEachChild<T extends FunctionlessNode>(
     const variableDecl = visitor(node.variableDecl);
     ensure(
       variableDecl,
-      isVariableStmt,
-      `VariableDecl in ${node.kind} must be a VariableDecl`
+      anyOf(isVariableStmt, isIdentifier),
+      `Initializer in ${node.kind} must be a VariableDecl or Identifier`
     );
 
     const expr = visitor(node.expr);
     ensure(expr, isExpr, `Expr in ${node.kind} must be an Expr`);
 
-    let body = visitor(node.body);
-    if (Array.isArray(body)) {
-      body = flatten(body);
-      ensureItemOf(body, isStmt, "Statements in BlockStmt must be Stmt nodes");
-      body = new BlockStmt(body);
-    } else {
-      ensure(body, isBlockStmt, `Body in ${node.kind} must be a BlockStmt`);
-    }
+    const body = blockStmtVisitor(node.body, visitor);
 
     return (
       isForInStmt(node)
         ? new ForInStmt(variableDecl, expr, body)
         : new ForOfStmt(variableDecl, expr, body)
     ) as T;
+  } else if (isForStmt(node)) {
+    const body = blockStmtVisitor(node.body, visitor);
```

Comment:
Why not just recurse instead of separating out logic into `blockStmtVisitor`?

Reply:
The visitor doesn't recurse, it handles each expr's children by passing it to the visitor and then validates the response. In the case of the block, you had allowed the normalization of a single or multiple exprs to block or a block itself being returned.

```ts
  let block = visitor(node);
  if (Array.isArray(block)) {
    block = flatten(block);
    ensureItemOf(block, isStmt, "Statements in BlockStmt must be Stmt nodes");
    return new BlockStmt(block);
  }
  ensure(block, isBlockStmt, `Body in ${node.kind} must be a BlockStmt`);
  return block;
```

If we recursed, the BlockStmt itself would never be passed to the visitor as the BlockStmt handler only passes the children down to the visitor, but the `For*` pass the BlockStmt itself down and accept multiple nodes as a result.

Reply:
Updated all cases of block handling to support returning an array of stmt or a blockstmt now. (it was inconsistent)

Reply:
Should we name it with the verb as the first word

```ts
visitBlockStmt(..)
```